diff --git a/lib/at91/at91.c b/lib/at91/at91.c
index 52b4642a02..62723a4cc6 100644
--- a/lib/at91/at91.c
+++ b/lib/at91/at91.c
@@ -16,7 +16,7 @@ void get_atmel_logo_info(vidinfo_t *info)
 	info->logo_width = ATMEL_LOGO_8BPP_WIDTH;
 	info->logo_height = ATMEL_LOGO_8BPP_HEIGHT;
 	info->logo_x_offset = ATMEL_LOGO_8BPP_X_OFFSET;
-	info->logo_y_offset = ATMEL_LOGO_8BPP_X_OFFSET;
+	info->logo_y_offset = ATMEL_LOGO_8BPP_Y_OFFSET;
 	info->logo_addr = (u_long)atmel_logo_8bpp;
 }
 
@@ -25,6 +25,6 @@ void get_microchip_logo_info(vidinfo_t *info)
 	info->logo_width = MICROCHIP_LOGO_8BPP_WIDTH;
 	info->logo_height = MICROCHIP_LOGO_8BPP_HEIGHT;
 	info->logo_x_offset = MICROCHIP_LOGO_8BPP_X_OFFSET;
-	info->logo_y_offset = MICROCHIP_LOGO_8BPP_X_OFFSET;
+	info->logo_y_offset = MICROCHIP_LOGO_8BPP_Y_OFFSET;
 	info->logo_addr = (u_long)microchip_logo_8bpp;
 }
diff --git a/lib/at91/microchip_logo_8bpp.h b/lib/at91/microchip_logo_8bpp.h
index 54daf530e1..b6101fc739 100644
--- a/lib/at91/microchip_logo_8bpp.h
+++ b/lib/at91/microchip_logo_8bpp.h
@@ -12,7 +12,7 @@
 #define MICROCHIP_LOGO_8BPP_HEIGHT	56
 
 #define MICROCHIP_LOGO_8BPP_X_OFFSET	0
-#define MCIROCHIP_LOGO_8BPP_Y_OFFSET	0
+#define MICROCHIP_LOGO_8BPP_Y_OFFSET	0
 
 /* Format: BMP 8BPP 240*60 */
 unsigned char microchip_logo_8bpp[] = {
