diff --git a/arch/arm/dts/Makefile b/arch/arm/dts/Makefile
index 21c6e9fa97..8ffcc31a60 100644
--- a/arch/arm/dts/Makefile
+++ b/arch/arm/dts/Makefile
@@ -343,6 +343,8 @@ dtb-$(CONFIG_TARGET_AT91SAM9X5EK) += \
 	at91sam9x25ek.dtb	\
 	at91sam9x35ek.dtb
 
+dtb-$(CONFIG_TARGET_RTH9580WF01) += rth9580wf01.dtb
+
 dtb-$(CONFIG_TARGET_AT91SAM9N12EK) += at91sam9n12ek.dtb
 
 dtb-$(CONFIG_TARGET_SAMA5D2_PTC_EK) += \
diff --git a/arch/arm/mach-at91/Kconfig b/arch/arm/mach-at91/Kconfig
index 62bef9ea76..71b13a9b5b 100644
--- a/arch/arm/mach-at91/Kconfig
+++ b/arch/arm/mach-at91/Kconfig
@@ -37,6 +37,13 @@ config TARGET_AT91SAM9261EK
 	bool "Atmel at91sam9261 reference board"
 	select CPU_ARM926EJS
 
+config TARGET_RTH9580WF01
+	bool "Resideo RTH9580WF01 thermostat board"
+	select CPU_ARM926EJS
+	select AT91SAM9X5
+	select BOARD_EARLY_INIT_F
+	select BOARD_LATE_INIT
+
 config TARGET_PM9261
 	bool "Ronetix pm9261 board"
 	select CPU_ARM926EJS
@@ -201,6 +208,7 @@ source "board/egnite/ethernut5/Kconfig"
 source "board/esd/meesc/Kconfig"
 source "board/l+g/vinco/Kconfig"
 source "board/mini-box/picosam9g45/Kconfig"
+source "board/resideo/rth9580wf01/Kconfig"
 source "board/ronetix/pm9261/Kconfig"
 source "board/ronetix/pm9263/Kconfig"
 source "board/ronetix/pm9g45/Kconfig"
diff --git a/board/resideo/rth9580wf01/Kconfig b/board/resideo/rth9580wf01/Kconfig
new file mode 100644
index 0000000000..015b7c9601
--- /dev/null
+++ b/board/resideo/rth9580wf01/Kconfig
@@ -0,0 +1,12 @@
+if TARGET_RTH9580WF01
+
+config SYS_BOARD
+	default "rth9580wf01"
+
+config SYS_VENDOR
+	default "resideo"
+
+config SYS_CONFIG_NAME
+	default "rth9580wf01"
+
+endif
diff --git a/board/resideo/rth9580wf01/MAINTAINERS b/board/resideo/rth9580wf01/MAINTAINERS
new file mode 100644
index 0000000000..0fccb46825
--- /dev/null
+++ b/board/resideo/rth9580wf01/MAINTAINERS
@@ -0,0 +1,6 @@
+RTH9580WF01 BOARD
+M:	John Clark <inindev@gmail.com>
+S:	Maintained
+F:	board/resideo/rth9580wf01/
+F:	include/configs/rth9580wf01.h
+F:	configs/rth9580wf01_spiflash_defconfig
diff --git a/board/resideo/rth9580wf01/Makefile b/board/resideo/rth9580wf01/Makefile
new file mode 100644
index 0000000000..ceb848b36f
--- /dev/null
+++ b/board/resideo/rth9580wf01/Makefile
@@ -0,0 +1,5 @@
+#
+# SPDX-License-Identifier:	GPL-2.0+
+#
+
+obj-y += rth9580wf01.o
diff --git a/board/resideo/rth9580wf01/rth9580wf01.c b/board/resideo/rth9580wf01/rth9580wf01.c
new file mode 100644
index 0000000000..d858d85168
--- /dev/null
+++ b/board/resideo/rth9580wf01/rth9580wf01.c
@@ -0,0 +1,136 @@
+/*
+ * Copyright (C) 2012 Atmel Corporation
+ *
+ * SPDX-License-Identifier:	GPL-2.0+
+ */
+
+#include <common.h>
+#include <dm.h>
+#include <asm/io.h>
+#include <asm/arch/at91sam9x5_matrix.h>
+#include <asm/arch/at91sam9_smc.h>
+#include <asm/arch/at91_common.h>
+#include <asm/arch/at91_rstc.h>
+#include <asm/arch/clk.h>
+#include <asm/arch/gpio.h>
+#include <debug_uart.h>
+#include <atmel_lcd.h>
+#include <version.h>
+#include <video.h>
+#ifdef CONFIG_DM_VIDEO
+#include <video_console.h>
+#endif
+
+DECLARE_GLOBAL_DATA_PTR;
+
+/* ------------------------------------------------------------------------- */
+/*
+ * miscelaneous platform dependent initialization
+ */
+
+#if defined(CONFIG_DM_VIDEO) && defined(CONFIG_ATMEL_HLCD)
+static int video_show_board_logo_info(void)
+{
+	struct video_priv *vid;
+	int i, j;
+	u32 len = 0;
+	char b;
+	char *p;
+	char buf[255];
+	char *corp = "Microchip Technology Inc.\n";
+	char temp[32];
+	struct udevice *dev, *con;
+	const char *s;
+	vidinfo_t logo_info;
+	int ret;
+
+	len += sprintf(&buf[len], "%s\n", U_BOOT_VERSION);
+	memcpy(&buf[len], corp, strlen(corp));
+	len += strlen(corp);
+	len += sprintf(&buf[len], "%s CPU at %s MHz\n", get_cpu_name(),
+			strmhz(temp, get_cpu_clk_rate()));
+
+	len += sprintf(&buf[len], "%d MB SDRAM, 16 MB Flash\n",
+		       CONFIG_SYS_SDRAM_SIZE >> 20);
+
+	ret = uclass_get_device(UCLASS_VIDEO, 0, &dev);
+	if (ret)
+		return ret;
+
+	get_microchip_logo_info(&logo_info);
+	p = (char*)logo_info.logo_addr;
+	i = *(int*)(logo_info.logo_addr + 10);
+	j = logo_info.logo_width * logo_info.logo_height + i - 1;
+	for( ; i<j; i++, j--) {
+		b = p[i];
+		p[i] = p[j];
+		p[j] = b;
+	}
+
+	vid = dev_get_uclass_priv(dev);
+	ret = video_bmp_display(dev, logo_info.logo_addr,
+				vid->xsize - logo_info.logo_width,
+				vid->ysize - logo_info.logo_height,
+				false);
+	if (ret)
+		return ret;
+
+	ret = uclass_get_device(UCLASS_VIDEO_CONSOLE, 0, &con);
+	if (ret)
+		return ret;
+
+	vidconsole_position_cursor(con, 0, (logo_info.logo_height + 8));
+	for (s = buf, i = 0; i < len; s++, i++)
+		vidconsole_put_char(con, *s);
+
+	return 0;
+}
+#endif
+
+#ifdef CONFIG_BOARD_LATE_INIT
+int board_late_init(void)
+{
+#ifdef CONFIG_DM_VIDEO
+	video_show_board_logo_info();
+#endif
+	return 0;
+}
+#endif
+
+#ifdef CONFIG_DEBUG_UART_BOARD_INIT
+void board_debug_uart_init(void)
+{
+	at91_seriald_hw_init();
+}
+#endif
+
+#ifdef CONFIG_BOARD_EARLY_INIT_F
+int board_early_init_f(void)
+{
+#ifdef CONFIG_DEBUG_UART
+	debug_uart_init();
+#endif
+	return 0;
+}
+#endif
+
+int board_init(void)
+{
+	/* arch number of AT91SAM9X5EK-Board */
+	gd->bd->bi_arch_number = MACH_TYPE_AT91SAM9X5EK;
+
+	/* adress of boot parameters */
+	gd->bd->bi_boot_params = CONFIG_SYS_SDRAM_BASE + 0x100;
+
+#if defined(CONFIG_USB_OHCI_NEW) || defined(CONFIG_USB_EHCI)
+	at91_uhp_hw_init();
+#endif
+	return 0;
+}
+
+int dram_init(void)
+{
+	gd->ram_size = get_ram_size((void *) CONFIG_SYS_SDRAM_BASE,
+					CONFIG_SYS_SDRAM_SIZE);
+	return 0;
+}
diff --git a/configs/rth9580wf01_spiflash_defconfig b/configs/rth9580wf01_spiflash_defconfig
new file mode 100644
index 0000000000..c97a0a2ce1
--- /dev/null
+++ b/configs/rth9580wf01_spiflash_defconfig
@@ -0,0 +1,69 @@
+CONFIG_ARM=y
+CONFIG_ARCH_AT91=y
+CONFIG_TARGET_RTH9580WF01=y
+CONFIG_SYS_MALLOC_F_LEN=0x2000
+CONFIG_DEFAULT_DEVICE_TREE="rth9580wf01"
+CONFIG_SYS_EXTRA_OPTIONS="AT91SAM9X5"
+CONFIG_SPI_BOOT=y
+# CONFIG_CONSOLE_MUX is not set
+CONFIG_SYS_CONSOLE_IS_IN_ENV=y
+CONFIG_HUSH_PARSER=y
+CONFIG_SYS_PROMPT="uboot> "
+CONFIG_CMD_BOOTZ=y
+# CONFIG_CMD_BOOTEFI is not set
+# CONFIG_CMD_ELF is not set
+# CONFIG_CMD_IMLS is not set
+CONFIG_CMD_MEMINFO=y
+# CONFIG_CMD_FLASH is not set
+CONFIG_CMD_SF=y
+CONFIG_CMD_USB=y
+# CONFIG_CMD_FPGA is not set
+CONFIG_CMD_GPIO=y
+# CONFIG_CMD_SETEXPR is not set
+# CONFIG_CMD_NET is not set
+CONFIG_CMD_DHCP=y
+CONFIG_CMD_PING=y
+# CONFIG_PARTITIONS is not set
+CONFIG_OF_CONTROL=y
+CONFIG_NET=y
+CONFIG_DM=y
+CONFIG_REGMAP=y
+CONFIG_SPL_REGMAP=y
+CONFIG_SYSCON=y
+CONFIG_SPL_SYSCON=y
+CONFIG_CLK=y
+CONFIG_CLK_AT91=y
+CONFIG_DM_GPIO=y
+CONFIG_AT91_GPIO=y
+# CONFIG_MMC is not set
+CONFIG_MTD=y
+CONFIG_DM_SPI_FLASH=y
+CONFIG_SPI_FLASH=y
+CONFIG_SPI_FLASH_ATMEL=y
+CONFIG_SPI_FLASH_ISSI=y
+CONFIG_SPI_FLASH_WINBOND=y
+CONFIG_SPI_FLASH_MTD=y
+CONFIG_DM_ETH=y
+CONFIG_MACB=y
+CONFIG_PINCTRL=y
+CONFIG_PINCTRL_AT91=y
+CONFIG_DM_SERIAL=y
+CONFIG_DEBUG_UART=y
+CONFIG_DEBUG_UART_ATMEL=y
+CONFIG_DEBUG_UART_BASE=0xfffff200
+CONFIG_DEBUG_UART_CLOCK=132000000
+CONFIG_DEBUG_UART_BOARD_INIT=y
+CONFIG_DEBUG_UART_ANNOUNCE=y
+CONFIG_ATMEL_USART=y
+CONFIG_DM_SPI=y
+CONFIG_ATMEL_SPI=y
+CONFIG_USB=y
+CONFIG_DM_USB=y
+CONFIG_USB_EHCI_HCD=y
+CONFIG_USB_STORAGE=y
+CONFIG_DM_VIDEO=y
+# CONFIG_VIDEO_BPP8 is not set
+# CONFIG_VIDEO_BPP32 is not set
+CONFIG_CONSOLE_ROTATION=y
+CONFIG_ATMEL_HLCD=y
+CONFIG_SMBIOS_MANUFACTURER="atmel"
diff --git a/drivers/mtd/spi/Kconfig b/drivers/mtd/spi/Kconfig
index 8eaf6814b6..8c5b8fa4bc 100644
--- a/drivers/mtd/spi/Kconfig
+++ b/drivers/mtd/spi/Kconfig
@@ -72,6 +72,11 @@ config SPI_FLASH_GIGADEVICE
 	help
 	  Add support for various GigaDevice SPI flash chips (GD25xxx)
 
+config SPI_FLASH_ISSI
+	bool "ISSI SPI flash support"
+	help
+	  Add support for various ISSI SPI flash chips (ISxxx)
+
 config SPI_FLASH_MACRONIX
 	bool "Macronix SPI flash support"
 	help
diff --git a/drivers/video/atmel_hlcdfb.c b/drivers/video/atmel_hlcdfb.c
index 89a6aaabb3..4e206685e4 100644
--- a/drivers/video/atmel_hlcdfb.c
+++ b/drivers/video/atmel_hlcdfb.c
@@ -489,6 +489,7 @@ static int atmel_hlcdc_probe(struct udevice *dev)
 	uc_priv->xsize = priv->timing.hactive.typ;
 	uc_priv->ysize = priv->timing.vactive.typ;
 	uc_priv->bpix = priv->vl_bpix;
+	uc_priv->rot = fdtdec_get_uint(gd->fdt_blob, dev->of_offset, "rotate", 0);
 
 	/* Enable flushing if we enabled dcache */
 	video_set_flush_dcache(dev, true);
diff --git a/include/configs/rth9580wf01.h b/include/configs/rth9580wf01.h
new file mode 100644
index 0000000000..738b16b544
--- /dev/null
+++ b/include/configs/rth9580wf01.h
@@ -0,0 +1,117 @@
+/*
+ * configuation settings for the rth9580wf01 board
+ *
+ * SPDX-License-Identifier:	GPL-2.0+
+ */
+
+#ifndef __RTH9580WF01_H__
+#define __RTH9580WF01_H__
+
+#include <asm/hardware.h>
+
+#define CONFIG_SYS_TEXT_BASE		0x20f50000
+
+/* ARM asynchronous clock */
+#define CONFIG_SYS_AT91_SLOW_CLOCK	32768
+#define CONFIG_SYS_AT91_MAIN_CLOCK	12000000	/* 12 MHz crystal */
+
+#define CONFIG_AT91SAM9X5EK
+
+#define CONFIG_CMDLINE_TAG		/* enable passing of ATAGs */
+#define CONFIG_SETUP_MEMORY_TAGS
+#define CONFIG_INITRD_TAG
+#define CONFIG_SKIP_LOWLEVEL_INIT
+
+/* general purpose I/O */
+#define CONFIG_ATMEL_LEGACY		/* required until (g)pio is fixed */
+
+/* LCD */
+#ifdef CONFIG_ATMEL_HLCD
+#define CONFIG_SYS_WHITE_ON_BLACK
+#endif
+
+/* BOOTP options */
+#define CONFIG_BOOTP_BOOTFILESIZE
+#define CONFIG_BOOTP_BOOTPATH
+#define CONFIG_BOOTP_GATEWAY
+#define CONFIG_BOOTP_HOSTNAME
+
+/* SDRAM */
+#define CONFIG_NR_DRAM_BANKS		1
+#define CONFIG_SYS_SDRAM_BASE		0x20000000
+#define CONFIG_SYS_SDRAM_SIZE		0x01000000	/* 16 megs */
+
+#define CONFIG_SYS_INIT_SP_ADDR \
+	(CONFIG_SYS_SDRAM_BASE + 16 * 1024 - GENERATED_GBL_DATA_SIZE)
+
+/* MTD */
+#define CONFIG_MTD_DEVICE
+#define CONFIG_CMD_MTDPARTS
+#define CONFIG_MTD_PARTITIONS
+#define CONFIG_RBTREE
+#define CONFIG_LZO
+
+/*
+ * SPI NOR
+ */
+#ifdef CONFIG_CMD_SF
+#define CONFIG_SF_DEFAULT_BUS		0
+#define CONFIG_SF_DEFAULT_CS		0
+#define CONFIG_SF_DEFAULT_MODE		SPI_MODE_0
+#define CONFIG_SF_DEFAULT_SPEED		50000000
+#endif
+
+/* USB */
+#ifdef CONFIG_CMD_USB
+#ifndef CONFIG_USB_EHCI
+#define CONFIG_USB_ATMEL
+#define CONFIG_USB_ATMEL_CLK_SEL_UPLL
+#define CONFIG_USB_OHCI_NEW
+#define CONFIG_SYS_USB_OHCI_CPU_INIT
+#define CONFIG_SYS_USB_OHCI_REGS_BASE		ATMEL_BASE_OHCI
+#define CONFIG_SYS_USB_OHCI_SLOT_NAME		"at91sam9x5"
+#define CONFIG_SYS_USB_OHCI_MAX_ROOT_PORTS	3
+#endif
+#endif
+
+#define CONFIG_SYS_LOAD_ADDR		0x20f50000	/* load address */
+
+#define CONFIG_SYS_MEMTEST_START	(CONFIG_SYS_SDRAM_BASE + SZ_32K)
+#define CONFIG_SYS_MEMTEST_END		(CONFIG_SYS_TEXT_BASE - SZ_1M)
+
+/* bootstrap + u-boot + env1 + env2 + linux in spi flash */
+#define CONFIG_ENV_IS_IN_SPI_FLASH
+#define CONFIG_SYS_REDUNDAND_ENVIRONMENT
+#define CONFIG_ENV_SECT_SIZE		0x1000
+#define CONFIG_ENV_SIZE			0x4000
+#define CONFIG_ENV_OFFSET		0x8000
+#define CONFIG_ENV_OFFSET_REDUND	(CONFIG_ENV_OFFSET + CONFIG_ENV_SIZE)
+#define CONFIG_ENV_SPI_BUS		0
+#define CONFIG_ENV_SPI_CS		0
+#define CONFIG_ENV_SPI_MODE		SPI_MODE_0
+#define CONFIG_ENV_SPI_MAX_HZ		50000000
+
+#define CONFIG_LZMA
+#define CONFIG_BOOTCOMMAND	"sf probe 0; " \
+				"sf read 0x20800000 0xc0000 0x400000; " \
+				"bootm 0x20800000 - 0x${fdtcontroladdr}"
+
+#define CONFIG_BAUDRATE		115200
+
+#define CONFIG_SYS_CBSIZE	256
+#define CONFIG_SYS_MAXARGS	16
+#define CONFIG_SYS_LONGHELP
+#define CONFIG_CMDLINE_EDITING
+#define CONFIG_AUTO_COMPLETE
+
+/*
+ * Size of malloc() pool
+ */
+#define CONFIG_SYS_MALLOC_LEN		(256 << 10)	/* reserve 256k for malloc() */
+
+#define CONFIG_SYS_MASTER_CLOCK		132096000
+#define CONFIG_SYS_AT91_PLLA		0x20c73f03
+#define CONFIG_SYS_MCKR			0x1301
+#define CONFIG_SYS_MCKR_CSS		0x1302
+
+#endif /* __RTH9580WF01_H__ */
