diff --git a/lib/at91/at91.c b/lib/at91/at91.c
index 048597690b..81d1be8530 100644
--- a/lib/at91/at91.c
+++ b/lib/at91/at91.c
@@ -15,7 +15,7 @@ void atmel_logo_info(vidinfo_t *info)
 	info->logo_width = ATMEL_LOGO_8BPP_WIDTH;
 	info->logo_height = ATMEL_LOGO_8BPP_HEIGHT;
 	info->logo_x_offset = ATMEL_LOGO_8BPP_X_OFFSET;
-	info->logo_y_offset = ATMEL_LOGO_8BPP_X_OFFSET;
+	info->logo_y_offset = ATMEL_LOGO_8BPP_Y_OFFSET;
 	info->logo_addr = (u_long)atmel_logo_8bpp;
 }
 
@@ -24,6 +24,6 @@ void microchip_logo_info(vidinfo_t *info)
 	info->logo_width = MICROCHIP_LOGO_8BPP_WIDTH;
 	info->logo_height = MICROCHIP_LOGO_8BPP_HEIGHT;
 	info->logo_x_offset = MICROCHIP_LOGO_8BPP_X_OFFSET;
-	info->logo_y_offset = MICROCHIP_LOGO_8BPP_X_OFFSET;
+	info->logo_y_offset = MICROCHIP_LOGO_8BPP_Y_OFFSET;
 	info->logo_addr = (u_long)microchip_logo_8bpp;
 }
diff --git a/lib/at91/microchip_logo_8bpp.h b/lib/at91/microchip_logo_8bpp.h
index a56f1607cb..7eb66958e9 100644
--- a/lib/at91/microchip_logo_8bpp.h
+++ b/lib/at91/microchip_logo_8bpp.h
@@ -11,7 +11,7 @@
 #define MICROCHIP_LOGO_8BPP_HEIGHT	56
 
 #define MICROCHIP_LOGO_8BPP_X_OFFSET	0
-#define MCIROCHIP_LOGO_8BPP_Y_OFFSET	0
+#define MICROCHIP_LOGO_8BPP_Y_OFFSET	0
 
 /* Format: BMP 8BPP 240*60 */
 unsigned char microchip_logo_8bpp[] = {

